core/src/Cache/CacheFallbackOperations.cs

```diff
@@ -214,10 +214,8 @@ internal class CacheFallbackOperations
                 {
                     RemoveEntriesWithMatchingId(environmentAliases, clientId, accountOrUserId, adalCache);
                 }
-                else
-                {
-                    RemoveEntriesWithMatchingName(environmentAliases, clientId, displayableId, adalCache);
-                }
+
+                RemoveEntriesWithMatchingName(environmentAliases, clientId, displayableId, adalCache);
```

Comment:
Why do we want to call RemoveEntriesWithMatchingName every time? This deletes based on displayableId and ignores accountId completely.

Reply:
this case is covered in UnifiedCache_Adalv3ToAdalV4ToMsal2MigrationIntegrationTest : 
if you get AT for res1 in  adal v3  than migrate to adal v4 and get token to res2 you will have next state of the cache:
for the same Account you have one entity in adal cache in adalv3 format (no client info)
another entity in adal v4 format (with client info).
So you have two Adal cache entities for the same account .

If you migrate to Msal and call get accounts api one Account object will be returned.
If you call Remove account it will remove only V4 adal entity , so next call to get accounts will return same account object (this time with id null)

So remove account api will not remove account .
Unless remove account api remove by displayble id also.

Reply:
We have been going back and forth on this. Last time we ended up saying we should not do both deletes as it have side effects. 

---
In reply to: [223143317](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1304#discussion_r223143317) [](ancestors = 223143317)

Reply:
not sure what side effects you mean here , In extremely rare case when few different accounts with same displaybleId/preferred_user_name were used in Adal, Msal remove Account api will remove SSO in adal cache format for both of them which is less evil that not working  removeAccount api for described scenario(which is quite possible scenario ) . If you want me to remove this change I can do it.

Reply:
I understand the scenario now, thank you for explaining Petro. Definitely needs to be in the code as a comment if we decided to go for it. Is this algorithm discussed with the other teams or is it only us that have this problem?

Reply:
Discussed with Olga, she implemented something similar

Reply:
Sounds good to me then. Note that there are more granular unit tests in LCacheFallbackOperationsTests.cs - we should add smth for this case

Reply:
Did you cover this with tests Petro?

---
In reply to: [224130526](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1304#discussion_r224130526) [](ancestors = 224130526)

Reply:
@bgavrilMS, are you ok with this change?

---
In reply to: [224675645](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1304#discussion_r224675645) [](ancestors = 224675645,224130526)

Reply:
@SomskaPe, are the tests in place for this?

---
In reply to: [226489008](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1304#discussion_r226489008) [](ancestors = 226489008,224675645,224130526)

Reply:
Yes, sounds good. Better to delete more than to not be able to remove accounts from cache...

---
In reply to: [226489100](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1304#discussion_r226489100) [](ancestors = 226489100,226489008,224675645,224130526)